src/util/apparmor.c

```diff
@@ -93,3 +105,446 @@ int bus_apparmor_dbus_supported(bool *supportedp) {
         *supportedp = supported;
         return 0;
 }
+
+static inline bool is_apparmor_enabled(BusAppArmorRegistry *registry) {
+        return registry->bustype ? true : false;
+}
+
+static int bus_apparmor_log(const char *fmt, ...) {
+        _c_cleanup_(c_freep) char *message = NULL;
+        va_list ap;
+        int r;
+
+        va_start(ap, fmt);
+        r = vasprintf(&message, fmt, ap);
+        va_end(ap);
+        if (r < 0)
+                return r;
+
+        /* XXX: we don't have access to any context, so can't find
+         * the right UID to use, follow dbus-daemon(1) and use our
+         * own. */
+        r = util_audit_log(UTIL_AUDIT_TYPE_AVC, message, getuid());
+        if (r)
+                return error_fold(r);
+
+        return 0;
+}
+
+/**
+ * bus_apparmor_registry_new() - create a new AppArmor registry
+ * @registryp:          pointer to the new registry
+ * @fallback_context:   fallback security context for queries against this registry
+ *
+ * Return: 0 on success, or a negative error code on failure.
+ */
+int bus_apparmor_registry_new(struct BusAppArmorRegistry **registryp, const char *fallback_context) {
+        _c_cleanup_(bus_apparmor_registry_unrefp) BusAppArmorRegistry *registry = NULL;
+        size_t n_fallback_context = strlen(fallback_context) + 1;
+
+        registry = malloc(sizeof(*registry) + n_fallback_context);
+        if (!registry)
+                return error_origin(-ENOMEM);
+
+        registry->n_refs = REF_INIT;
+        registry->fallback_context = (char *)(registry + 1);
+        memcpy((char *)registry->fallback_context, fallback_context, n_fallback_context);
+
+        *registryp = registry;
+        registry = NULL;
+        return 0;
+}
+
+static void bus_apparmor_registry_free(_Atomic unsigned long *n_refs, void *userdata) {
+        BusAppArmorRegistry *registry = c_container_of(n_refs, BusAppArmorRegistry, n_refs);
+
+        free(registry->bustype);
+        free(registry);
+}
+
+BusAppArmorRegistry *bus_apparmor_registry_ref(BusAppArmorRegistry *registry) {
+        if (registry)
+                ref_inc(&registry->n_refs);
+
+        return registry;
+}
+
+BusAppArmorRegistry *bus_apparmor_registry_unref(BusAppArmorRegistry *registry) {
+        if (registry)
+                ref_dec(&registry->n_refs, bus_apparmor_registry_free, NULL);
+
+        return NULL;
+}
+
+int bus_apparmor_set_bus_type(BusAppArmorRegistry *registry, const char* bustype) {
+        c_assert(!registry->bustype);
+
+        registry->bustype = strdup(bustype);
+        if (!registry->bustype)
+                return error_origin(-ENOMEM);
+        return 0;
+}
+
+static bool is_unconfined(const char *context) {
+        return !strcmp(context, "unconfined");
```

Comment:
dbus-daemon checks for the mode to be equal to `unconfined`, or unset and the label equal to `unconfined`. Any particular reason this is different?